README.rst
disqus-python
~~~~~~~~~~~~~

.. image:: https://travis-ci.org/disqus/disqus-python.svg?branch=master
    :target: https://travis-ci.org/disqus/disqus-python

Let's start with installing the API:

	pip install disqus-python

Use the API by instantiating it, and then calling the method through dotted notation chaining::

	from disqusapi import DisqusAPI
	disqus = DisqusAPI(secret_key, public_key)
	for result in disqus.get('trends.listThreads'):
	    print result

Parameters (including the ability to override version, api_secret, and format) are passed as keyword arguments to the resource call::

	disqus.get('posts.details', post=1, version='3.0')

Paginating through endpoints is easy as well::

	from disqusapi import Paginator
	paginator = Paginator(api.get, 'trends.listThreads', forum='disqus')
	for result in paginator:
	    print result

	# pull in a maximum of 500 results (this limit param differs from the endpoint's limit param)
	for result in paginator(limit=500):
	    print result

Documentation on all methods, as well as general API usage can be found at https://disqus.com/api/docs/


CHANGES
0.5.0

* Added tox support.
* New invocation api.<method>(<endpoint>, <params>) where endpoint
  is a string. This to add compatibility when the endpoint contains
  reserved words.
* Python 3 support.
* Removed simplejson dependency
* Travis CI support.
* Added socket timeout.
* Deprecated the `setSecretKey` and `setKey` methods in favor of the `secret_key` attribute.
* Deprecated the `key` property in favor of the `secret_key` attribute.
* Deprecated the `setPublicKey` method in favor of directly setting the `public_key` attribute.
* Deprecated the `setFormat` method in favor of directly setting the `format` attribute.
* Deprecated the `setVersion` method in favor of directly setting the `version` attribute.
* Deprecated the `setTimeout` method in favor of directly setting the `timeout` attribute.
* Request gzipped responses from API.

0.4.1

* Fixed behavior when api_secret or api_public are not set.

0.4.0

* Removed signed requests (Disqus has deprecated them).
* Change all endpoints to use SSL.

0.3.2

* Added support for undefined interfaces.
* Added support for mapping error codes to different exceptions.

0.3.1

* Fixed an issue with GET requests and the normalized request string.

0.3.0

* Added signed requests (you must pass public_key to DisqusAPI).
* Added support for OAuth (via access_token in signed requests).
* Moved Paginator into disqusapi.paginator


disqusapi/tests.py
import mock
import os
import socket

import disqusapi
from disqusapi.compat import xrange
from disqusapi.tests_compat import TestCase

extra_interface = {
    "reserved": {
        "global": {
            "word": {
                "method": "GET",
                "required": [
                    "text",
                ],
                "formats": [
                    "json",
                ],
            }
        }
    }
}


extra_interface = {
    "reserved": {
        "global": {
            "word": {
                "method": "GET",
                "required": [
                    "text",
                ],
                "formats": [
                    "json",
                ],
            }
        }
    }
}


def requires(*env_vars):
    def wrapped(func):
        for k in env_vars:
            if not os.environ.get(k):
                return
        return func
    return wrapped


def iter_results():
    for n in xrange(11):
        yield disqusapi.Result(
            response=[n] * 10,
            cursor={
                'id': n,
                'more': n < 10,
            },
        )


class MockResponse(object):
    def __init__(self, body, status=200):
        self.body = body
        self.status = status

    def read(self):
        return self.body


class DisqusAPITest(TestCase):
    API_SECRET = 'b' * 64
    API_PUBLIC = 'c' * 64
    HOST = os.environ.get('DISQUS_API_HOST', disqusapi.HOST)

    def setUp(self):
        disqusapi.HOST = self.HOST

    def test_setKey(self):
        api = disqusapi.DisqusAPI('a', 'c')
        self.assertEquals(api.secret_key, 'a')
        api.setKey('b')
        self.assertEquals(api.secret_key, 'b')

    def test_setSecretKey(self):
        api = disqusapi.DisqusAPI('a', 'c')
        self.assertEquals(api.secret_key, 'a')
        api.setSecretKey('b')
        self.assertEquals(api.secret_key, 'b')

    def test_setPublicKey(self):
        api = disqusapi.DisqusAPI('a', 'c')
        self.assertEquals(api.public_key, 'c')
        api.setPublicKey('b')
        self.assertEquals(api.public_key, 'b')

    def test_setFormat(self):
        api = disqusapi.DisqusAPI()
        self.assertEquals(api.format, 'json')
        api.setFormat('jsonp')
        self.assertEquals(api.format, 'jsonp')

    def test_setVersion(self):
        api = disqusapi.DisqusAPI()
        self.assertEquals(api.version, '3.0')
        api.setVersion('3.1')
        self.assertEquals(api.version, '3.1')

    def test_setTimeout(self):
        api = disqusapi.DisqusAPI()
        self.assertEquals(api.timeout, socket.getdefaulttimeout())
        api = disqusapi.DisqusAPI(timeout=30)
        self.assertEquals(api.timeout, 30)
        api.setTimeout(60)
        self.assertEquals(api.timeout, 60)

    def test_paginator(self):
        api = disqusapi.DisqusAPI(self.API_SECRET, self.API_PUBLIC)
        with mock.patch('disqusapi.Resource._request') as _request:
            iterator = iter_results()
            _request.return_value = next(iterator)
            paginator = disqusapi.Paginator(api, 'posts.list', forum='disqus')
            n = 0
            for n, result in enumerate(paginator(limit=100)):
                if n % 10 == 0:
                    next(iterator)
        self.assertEquals(n, 99)

    def test_paginator_legacy(self):
        api = disqusapi.DisqusAPI(self.API_SECRET, self.API_PUBLIC)
        with mock.patch('disqusapi.Resource._request') as _request:
            iterator = iter_results()
            _request.return_value = next(iterator)
            paginator = disqusapi.Paginator(api.posts.list, forum='disqus')
            n = 0
            for n, result in enumerate(paginator(limit=100)):
                if n % 10 == 0:
                    next(iterator)
        self.assertEquals(n, 99)

    def test_endpoint(self):
        api = disqusapi.DisqusAPI(self.API_SECRET, self.API_PUBLIC)
        with mock.patch('disqusapi.Resource._request') as _request:
            iterator = iter_results()
            _request.return_value = next(iterator)
            response1 = api.posts.list(forum='disqus')

        with mock.patch('disqusapi.Resource._request') as _request:
            iterator = iter_results()
            _request.return_value = next(iterator)
            response2 = api.get('posts.list', forum='disqus')

        self.assertEquals(len(response1), len(response2))

    def test_update_interface_legacy(self):
        api = disqusapi.DisqusAPI(self.API_SECRET, self.API_PUBLIC)
        with self.assertRaises(disqusapi.InterfaceNotDefined):
            api.interface.update(extra_interface)

    def test_invalid_method(self):
        api = disqusapi.DisqusAPI(self.API_SECRET, self.API_PUBLIC)
        with self.assertRaises(disqusapi.InvalidHTTPMethod):
            api.get('posts.list', method='lol', forum='disqus')

    def test_update_interface(self):
        api = disqusapi.DisqusAPI(self.API_SECRET, self.API_PUBLIC)
        api.update_interface(extra_interface)

if __name__ == '__main__':
    import unittest
    unittest.main()


runtests.py
import unittest

from disqusapi import DisqusAPI, APIError

class DisqusAPITest(unittest.TestCase):
    API_SECRET = 'b'*64

    def test_setKey(self):
        api = DisqusAPI('a')
        self.assertEquals(api.key, 'a')
        api.setKey('b')
        self.assertEquals(api.key, 'b')

    def test_setFormat(self):
        api = DisqusAPI()
        self.assertEquals(api.format, 'json')
        api.setFormat('jsonp')
        self.assertEquals(api.format, 'jsonp')
    
    def test_setVersion(self):
        api = DisqusAPI()
        self.assertEquals(api.version, '3.0')
        api.setVersion('3.1')
        self.assertEquals(api.version, '3.1')
    
    def test_users_listActivity(self):
        api = DisqusAPI(self.API_SECRET)
        self.assertRaises(APIError, api.users.listActivity, foo='bar')

if __name__ == '__main__':
    unittest.main()

disqusapi/utils.py
import binascii
import hashlib
import hmac
from disqusapi import compat
from disqusapi.compat import urllib_parse as urlparse


def build_interfaces_by_method(interfaces):
    """
    Create new dictionary from INTERFACES hashed by method then
    the endpoints name. For use when using the disqusapi by the
    method interface instead of the endpoint interface. For
    instance:

    'blacklists': {
        'add': {
            'formats': ['json', 'jsonp'],
            'method': 'POST',
            'required': ['forum']
        }
    }

    is translated to:

    'POST': {
        'blacklists.add': {
            'formats': ['json', 'jsonp'],
            'method': 'POST',
            'required': ['forum']
    }
    """
    def traverse(block, parts):
        try:
            method = block['method'].lower()
        except KeyError:
            for k, v in compat.iteritems(block):
                traverse(v, parts + [k])
        else:
            path = '.'.join(parts)
            try:
                methods[method]
            except KeyError:
                methods[method] = {}
            methods[method][path] = block
    methods = {}
    for key, val in compat.iteritems(interfaces):
        traverse(val, [key])
    return methods


def get_normalized_params(params):
    """
    Given a list of (k, v) parameters, returns
    a sorted, encoded normalized param
    """
    return urlparse.urlencode(sorted(params))


def get_normalized_request_string(method, url, nonce, params, ext='', body_hash=None):
    """
    Returns a normalized request string as described iN OAuth2 MAC spec.

    http://tools.ietf.org/html/draft-ietf-oauth-v2-http-mac-00#section-3.3.1
    """
    urlparts = urlparse.urlparse(url)
    if urlparts.query:
        norm_url = '%s?%s' % (urlparts.path, urlparts.query)
    elif params:
        norm_url = '%s?%s' % (urlparts.path, get_normalized_params(params))
    else:
        norm_url = urlparts.path

    if not body_hash:
        body_hash = get_body_hash(params)

    port = urlparts.port
    if not port:
        assert urlparts.scheme in ('http', 'https')

        if urlparts.scheme == 'http':
            port = 80
        elif urlparts.scheme == 'https':
            port = 443

    output = [nonce, method.upper(), norm_url, urlparts.hostname, port, body_hash, ext, '']

    return '\n'.join(map(str, output))


def get_body_hash(params):
    """
    Returns BASE64 ( HASH (text) ) as described in OAuth2 MAC spec.

    http://tools.ietf.org/html/draft-ietf-oauth-v2-http-mac-00#section-3.2
    """
    norm_params = get_normalized_params(params)

    return binascii.b2a_base64(hashlib.sha1(norm_params).digest())[:-1]


def get_mac_signature(api_secret, norm_request_string):
    """
    Returns HMAC-SHA1 (api secret, normalized request string)
    """
    hashed = hmac.new(str(api_secret), norm_request_string, hashlib.sha1)
    return binascii.b2a_base64(hashed.digest())[:-1]


disqusapi/paginator.py
class Paginator(object):
    """
    Paginate through all entries:

    >>> paginator = Paginator(api, 'trends.listThreads', forum='disqus')
    >>> for result in paginator:
    >>>     print result

    Paginate only up to a number of entries:

    >>> for result in paginator(limit=500):
    >>>     print result
    """

    def __init__(self, *args, **params):
        from disqusapi import InterfaceNotDefined
        if len(args) == 2:
            self.method = args[0]
            self.endpoint = args[1]
        elif len(args) == 1:
            self.method = None
            self.endpoint = args[0]
        else:
            raise InterfaceNotDefined
        self.params = params

    def __iter__(self):
        for result in self():
            yield result

    def __call__(self, limit=None):
        params = self.params.copy()
        num = 0
        more = True
        while more and (not limit or num < limit):
            if self.method:
                results = self.method(self.endpoint, **params)
            else:
                results = self.endpoint(**params)
            for result in results:
                if limit and num >= limit:
                    break
                num += 1
                yield result

            if results.cursor:
                more = results.cursor['more']
                params['cursor'] = results.cursor['id']
            else:
                more = False


disqusapi/tests_compat.py
import unittest
import sys

if sys.version_info < (2, 7):
    # Stolen from unittest2
    import re

    class _AssertRaisesBaseContext(object):

        def __init__(self, expected, test_case, callable_obj=None,
                     expected_regex=None):
            self.expected = expected
            self.failureException = test_case.failureException
            if callable_obj is not None:
                try:
                    self.obj_name = callable_obj.__name__
                except AttributeError:
                    self.obj_name = str(callable_obj)
            else:
                self.obj_name = None
            if isinstance(expected_regex, basestring):  # NOQA
                expected_regex = re.compile(expected_regex)
            self.expected_regex = expected_regex

    class _AssertRaisesContext(_AssertRaisesBaseContext):
        """A context manager used to implement TestCase.assertRaises* methods."""

        def __enter__(self):
            return self

        def __exit__(self, exc_type, exc_value, tb):
            if exc_type is None:
                try:
                    exc_name = self.expected.__name__
                except AttributeError:
                    exc_name = str(self.expected)
                raise self.failureException(
                    "%s not raised" % (exc_name,))
            if not issubclass(exc_type, self.expected):
                # let unexpected exceptions pass through
                return False
            self.exception = exc_value  # store for later retrieval
            if self.expected_regex is None:
                return True

            expected_regex = self.expected_regex
            if not expected_regex.search(str(exc_value)):
                raise self.failureException(
                    '%r does not match %r' %
                    (expected_regex.pattern, str(exc_value)))
            return True

    class TestCase(unittest.TestCase):
        def assertRaises(self, excClass, callableObj=None, *args, **kwargs):
            """Fail unless an exception of class excClass is thrown
               by callableObj when invoked with arguments args and keyword
               arguments kwargs. If a different type of exception is
               thrown, it will not be caught, and the test case will be
               deemed to have suffered an error, exactly as for an
               unexpected exception.

               If called with callableObj omitted or None, will return a
               context object used like this::

                    with self.assertRaises(SomeException):
                        do_something()

               The context manager keeps a reference to the exception as
               the 'exception' attribute. This allows you to inspect the
               exception after the assertion::

                   with self.assertRaises(SomeException) as cm:
                       do_something()
                   the_exception = cm.exception
                   self.assertEqual(the_exception.error_code, 3)
            """
            if callableObj is None:
                return _AssertRaisesContext(excClass, self)
            try:
                callableObj(*args, **kwargs)
            except excClass:
                return

            if hasattr(excClass, '__name__'):
                excName = excClass.__name__
            else:
                excName = str(excClass)
            raise self.failureException("%s not raised" % excName)
else:
    TestCase = unittest.TestCase


disqusapi/__init__.py
"""
disqus-python
~~~~~~~~~~~~~

disqus = DisqusAPI(api_secret=secret_key)
disqus.get('trends.listThreads')

"""
try:
    __version__ = __import__('pkg_resources') \
        .get_distribution('disqusapi').version
except Exception:  # pragma: no cover
    __version__ = 'unknown'

import re
import zlib
import os.path
import warnings
import socket

try:
    import simplejson as json
except ImportError:
    import json

from disqusapi.paginator import Paginator
from disqusapi import compat
from disqusapi.compat import http_client as httplib
from disqusapi.compat import urllib_parse as urllib
from disqusapi.utils import build_interfaces_by_method

__all__ = ['DisqusAPI', 'Paginator']

with open(os.path.join(os.path.dirname(__file__), 'interfaces.json')) as fp:
    INTERFACES = json.load(fp)

HOST = 'disqus.com'

CHARSET_RE = re.compile(r'charset=(\S+)')
DEFAULT_ENCODING = 'utf-8'


class InterfaceNotDefined(NotImplementedError):
    pass


class InvalidHTTPMethod(TypeError):
    def __init__(self, message):
        self.message = message

    def __str__(self):
        return "expected 'GET' or 'POST', got: %r" % self.message


class FormattingError(ValueError):
    pass


class APIError(Exception):
    def __init__(self, code, message):
        self.code = code
        self.message = message

    def __str__(self):
        return '%s: %s' % (self.code, self.message)


class InvalidAccessToken(APIError):
    pass

ERROR_MAP = {
    18: InvalidAccessToken,
}


class Result(object):
    def __init__(self, response, cursor=None):
        self.response = response
        self.cursor = cursor or {}

    def __repr__(self):
        return '<%s: %s>' % (self.__class__.__name__, repr(self.response))

    def __iter__(self):
        for r in self.response:
            yield r

    def __len__(self):
        return len(self.response)

    def __getslice__(self, i, j):
        return list.__getslice__(self.response, i, j)

    def __getitem__(self, key):
        return list.__getitem__(self.response, key)

    def __contains__(self, key):
        return list.__contains__(self.response, key)


class Resource(object):
    def __init__(self, api, interfaces=INTERFACES, node=None, tree=()):
        self.api = api
        self.node = node
        self.interfaces = interfaces
        if node:
            tree = tree + (node,)
        self.tree = tree
        self.interfaces_by_method = {}

    def update_interface(self, interface):
        raise NotImplemented

    def __getattr__(self, attr):
        if attr in getattr(self, '__dict__'):
            return getattr(self, attr)
        if attr == 'interface':
            raise InterfaceNotDefined('You must use ``update_interface`` now.')
        interface = {}
        try:
            interface = self.interfaces[attr]
        except KeyError:
            try:
                interface = self.interfaces_by_method[attr]
            except KeyError:
                pass
        return Resource(self.api, interface, attr, self.tree)

    def __call__(self, endpoint=None, **kwargs):
        return self._request(endpoint, **kwargs)

    def _request(self, endpoint=None, **kwargs):
        if endpoint is not None:
            # Handle undefined interfaces
            resource = self.interfaces.get(endpoint, {})
            endpoint = endpoint.replace('.', '/')
        else:
            resource = self.interfaces
            endpoint = '/'.join(self.tree)
        for k in resource.get('required', []):
            if k not in (x.split(':')[0] for x in compat.iterkeys(kwargs)):
                raise ValueError('Missing required argument: %s' % k)

        method = kwargs.pop('method', resource.get('method'))

        if not method:
            raise InterfaceNotDefined(
                'Interface is not defined, you must pass ``method`` (HTTP Method).')

        method = method.upper()
        if method not in ('GET', 'POST'):
            raise InvalidHTTPMethod(method)

        api = self.api

        version = kwargs.pop('version', api.version)
        format = kwargs.pop('format', api.format)
        formatter, formatter_error = api.formats[format]

        path = '/api/%s/%s.%s' % (version, endpoint, format)

        if 'api_secret' not in kwargs and api.secret_key:
            kwargs['api_secret'] = api.secret_key
        if 'api_public' not in kwargs and api.public_key:
            kwargs['api_key'] = api.public_key

        # We need to ensure this is a list so that
        # multiple values for a key work
        params = []
        for k, v in compat.iteritems(kwargs):
            if isinstance(v, (list, tuple)):
                for val in v:
                    params.append((k, val))
            else:
                params.append((k, v))

        headers = {
            'User-Agent': 'disqus-python/%s' % __version__,
            'Accept-Encoding': 'gzip',
        }

        if method == 'GET':
            path = '%s?%s' % (path, urllib.urlencode(params))
            data = ''
        else:
            data = urllib.urlencode(params)

        conn = httplib.HTTPSConnection(HOST, timeout=api.timeout)
        conn.request(method, path, data, headers)
        response = conn.getresponse()

        try:
            body = response.read()
        finally:
            # Close connection
            conn.close()

        if response.getheader('Content-Encoding') == 'gzip':
            # See: http://stackoverflow.com/a/2424549
            body = zlib.decompress(body, 16 + zlib.MAX_WBITS)

        # Determine the encoding of the response and respect
        # the Content-Type header, but default back to utf-8
        content_type = response.getheader('Content-Type')
        if content_type is None:
            encoding = DEFAULT_ENCODING
        else:
            try:
                encoding = CHARSET_RE.search(content_type).group(1)
            except AttributeError:
                encoding = DEFAULT_ENCODING

        body = body.decode(encoding)

        try:
            # Coerce response to Python
            data = formatter(body)
        except formatter_error:
            raise FormattingError(body)

        if response.status != 200:
            raise ERROR_MAP.get(data['code'], APIError)(data['code'], data['response'])

        if isinstance(data['response'], list):
            return Result(data['response'], data.get('cursor'))
        return data['response']


class DisqusAPI(Resource):
    formats = {
        'json': (json.loads, ValueError),
    }

    def __init__(self, secret_key=None, public_key=None, format='json', version='3.0',
                 timeout=None, interfaces=INTERFACES, **kwargs):
        self.secret_key = secret_key
        self.public_key = public_key
        if not public_key:
            warnings.warn('You should pass ``public_key`` in addition to your secret key.')
        self.format = format
        self.version = version
        self.timeout = timeout or socket.getdefaulttimeout()
        self.interfaces = interfaces
        self.interfaces_by_method = build_interfaces_by_method(self.interfaces)
        super(DisqusAPI, self).__init__(self)

    @property
    def key(self):
        warnings.warn(
            "'key' is deprecated in favor of directly accessing the `secret_key` attribute",
            DeprecationWarning)
        return self.secret_key

    def setSecretKey(self, key):
        warnings.warn(
            "'setSecretKey' is deprecated in favor of directly setting the 'secret_key' attribute",
            DeprecationWarning)
        self.secret_key = key

    def setKey(self, key):
        warnings.warn(
            "'setKey' is deprecated in favor of directly setting the 'secret_key' attribute",
            DeprecationWarning)
        self.secret_key = key

    def setPublicKey(self, key):
        warnings.warn(
            "'setPublicKey' is deprecated in favor of directly setting the 'public_key' attribute",
            DeprecationWarning)
        self.public_key = key

    def setFormat(self, format):
        warnings.warn(
            "'setFormat' is deprecated in favor of directly setting the 'format' attribute",
            DeprecationWarning)
        self.format = format

    def setVersion(self, version):
        warnings.warn(
            "'setVersion' is deprecated in favor of directly setting the 'version' attribute",
            DeprecationWarning)
        self.version = version

    def setTimeout(self, timeout):
        warnings.warn(
            "'setTimeout' is deprecated in favor of directly setting the 'timeout' attribute",
            DeprecationWarning)
        self.timeout = timeout

    def update_interface(self, new_interface):
        self.interfaces.update(new_interface)
        self.interfaces_by_method = build_interfaces_by_method(self.interfaces)
